### PR TITLE
Setup Playwright test suite

### PR DESCRIPTION
**Pull request by petersalomonsen**
## Overview

This PR sets up a complete Playwright testing infrastructure for the Next.js treasury app, including comprehensive E2E tests for the intents deposit functionality and test automation tooling.

https://github.com/user-attachments/assets/9ad9aea4-3be0-435f-b20d-ab290dad4cfb

https://github.com/user-attachments/assets/890c364a-3e78-4fb5-b816-1e980c302903

## Key Achievements

✅ **Complete Playwright setup** with DevContainer, CI/CD, and browser support
✅ **8 comprehensive E2E tests** (1 dashboard + 7 intents deposit UI)
✅ **Video merger tool** for creating test demo videos
✅ **GitHub Actions CI/CD** workflow with automated testing
✅ **Browser compatibility** handling (Chromium fully supported, WebKit/Firefox skip clipboard tests)

### Dashboard Tests (1 test)
- ✅ Open dashboard for a valid DAO ID

### Intents Deposit UI Tests (7 tests)
- ✅ Display deposit button in Total Balance card
- ✅ Deposit button dropdown and Sputnik DAO navigation with clipboard verification
- ✅ Navigate between Sputnik DAO and NEAR Intents deposit pages
- ✅ Display QR code in Sputnik DAO deposit page
- ✅ Verify deposit addresses and QR codes for all assets and networks (10 random assets)
- ✅ Search for an asset and select it
- ✅ Display human-readable blockchain names in network selection (10 random assets)

**Browser Support:**
- **Chromium**: All 8 tests passing ✅
- **Firefox/WebKit**: Dashboard test passing; intents tests skipped (clipboard API not supported)

### 1. Test Infrastructure Setup

**Playwright Configuration** (`playwright.config.js`)
- Configured Chromium, Firefox, and WebKit browsers
- Automatic Next.js dev server startup via webServer config
- Video recording: on locally, retain-on-failure in CI
- Proper video object format to enable test-title.txt generation
- Screenshot capture on failures
- Retries and parallelization for CI

**DevContainer Enhancements**
- Pre-configured Node.js environment with all dependencies
- Remote desktop via noVNC for visual test debugging
- All Playwright browsers pre-installed
- Forwarded desktop ports for accessibility

**GitHub Actions CI/CD** (`.github/workflows/playwright.yml`)
- Runs on PRs, pushes to main, and manual dispatch
- Node.js 20 with npm caching
- Installs Playwright browsers with system dependencies
- Uploads test artifacts on failure
- 60-minute timeout with retry logic

### 2. Intents Deposit UI Tests

**Port from Legacy BOS Project**
Ported comprehensive intents deposit tests from the legacy `neardevhub-treasury-dashboard` project with significant improvements:

**Key Improvements:**
- ✅ **Backend API Integration**: Fixed metadata fetching to use backend API (`https://api.intents.near.social`) instead of unreliable external API
- ✅ **Network Name Mapping**: Uses `chainName` field from backend for consistent network identification
- ✅ **Improved Selectors**: Enhanced address card and modal selectors for reliability
- ✅ **QR Code Validation**: Comprehensive QR code reading and address verification using jimp and jsqr
- ✅ **Dynamic Asset Testing**: Tests 10 random assets per run for broad coverage
- ✅ **Clipboard Verification**: Tests copy-to-clipboard functionality with proper API permissions

**Test Utilities** (`playwright-tests/util/test.js`)
- Custom test fixture with treasury-specific options (daoAccount, instanceAccount, etc.)
- Automatic test-title.txt generation for video merger
- Overlay message utilities for visual test feedback
- Web3Icons integration for asset verification

### 3. Video Merger Tool

**Purpose**: Create professional demo videos from test recordings

**Script** (`scripts/playwright-video-merger.js`)
- Finds all test video recordings in `test-results/`
- Overlays test titles at bottom of each video
- Adds 1-second freeze frames between tests
- Merges everything into `final_output.mp4`

**Usage:**
```bash
npm run test:video-merge
```

**Features:**
- Reads test titles from auto-generated `test-title.txt` files
- Falls back to directory name if title file missing
- Wraps long titles for better readability
- JavaScript-based (replaced problematic bash script)

### 4. Documentation

**README.md Updates**
- Added comprehensive Testing section
- Documented all E2E test commands
- Video merger usage instructions
- Requirements and features

**Test Plan Organization**
- Moved `PLAYWRIGHT_TEST_PLAN.md` to `docs/ai-implementation-plans/`
- Better project structure for AI implementation documentation

### 5. Browser Compatibility

**Clipboard API Limitations**
- Clipboard permissions only supported in Chromium
- WebKit/Firefox tests gracefully skipped with clear messaging
- All clipboard-dependent tests confined to Chromium

## Test Scripts

```bash
npm run test:e2e              # Run all tests
npm run test:e2e:ui           # Run tests in UI mode (requires remote desktop)
npm run test:e2e:headed       # Run tests in headed mode
npm run test:e2e:debug        # Run tests in debug mode
npm run test:e2e:report       # View test report
npm run test:video-merge      # Create merged demo video
```

### Metadata API Integration
The intents deposit tests use the Intents backend API for reliable token metadata:
- **Endpoint**: `https://api.intents.near.social/metadata/extended`
- **Features**: Includes decimals, blockchain, symbol, chainName, price, icon
- **Benefits**: Consistent data, no external API failures, proper network mapping

### Test Data Strategy
Tests use dynamic asset selection:
- Randomly selects 10 assets per test run from 91 available tokens
- Ensures broad coverage across multiple runs
- Tests various blockchain networks (Ethereum, Near, Solana, Tron, etc.)

### Video Recording Configuration
```javascript
video: {
  mode: process.env.CI ? 'retain-on-failure' : 'on',
  size: { width: 1280, height: 800 },
}
```
This format ensures `page.video()` returns a valid object, enabling test-title.txt generation.

### New Files
- `playwright-tests/tests/intents/intents-deposit-ui.spec.js` - 7 comprehensive intents tests
- `scripts/playwright-video-merger.js` - Video merger tool
- `docs/ai-implementation-plans/PLAYWRIGHT_TEST_PLAN.md` - Moved from root

### Modified Files
- `playwright.config.js` - Updated video config format
- `playwright-tests/tests/dashboard/open-dashboard.spec.js` - Uses custom test utility
- `playwright-tests/util/test.js` - Added test-title.txt generation
- `package.json` - Added `test:video-merge` script
- `README.md` - Added comprehensive testing documentation
- `.gitignore` - Added video merger output directories

## Known Limitations

1. **Clipboard API**: WebKit and Firefox don't support clipboard permissions, so intents deposit tests only run on Chromium
2. **Random Asset Selection**: Tests use random assets, so specific edge cases may not be hit on every run

## Next Steps

Future enhancements (separate PRs):
1. Add transaction tests with near-sandbox
2. Set up indexer API test instance
3. Port additional tests from original BOS project
4. Add test data fixtures for deterministic testing
5. Expand dashboard test coverage

## Related Issue

Addresses #12

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by petersalomonsen** [comment]
@race-of-sloths include

**Review by Megha-Dev-19 — APPROVED** [review]
@race-of-sloths score 5

File: docs/ai-implementation-plans/PLAYWRIGHT_TEST_PLAN.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
+# Playwright Test Suite Setup Plan
+
+## Overview
```
[the remark this answers is not in the document]
Comment (petersalomonsen) [reply docs/ai-implementation-plans/PLAYWRIGHT_TEST_PLAN.md]: yes. since the plan is not completely finalized yet. still going to implement the near-sandbox support, which is part of this plan.

File: .devcontainer/post-create.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+# Install project dependencies
+npm install
+
+# Install Playwright system dependencies and browsers
+npx playwright install-deps
+npx playwright install
+
+# Install Claude Code
+npm install -g @anthropic-ai/claude-code
```
[the remark this answers is not in the document]
Comment (petersalomonsen) [reply .devcontainer/post-create.sh]: It's only for the devcontainer, so not installed if people want to develop locally. It's optional to use it, but very convenient if you use it a lot like me.
